src/tape_drivers/freebsd/cam/camtape_tc.c

```diff
@@ -0,0 +1,4206 @@
+	return rc;
+}
+
+/**
+ * Write attribute
+ * @param device a pointer to the camtape backend
+ * @param part partition to write attribute
+ * @param buf pointer to the attribute buffer as defined in SPC-4 7.4.1
+ * 	  "Attribute Format". The header for the parameter list is created based
+ * 	  on the size of the attribute buffer
+ * @param size length of the attribute buffer
+ * @return 0 on success or a negative value on error
+ */
+int camtape_write_attribute(void *device, const tape_partition_t part, const unsigned char *buf,
+							const size_t size)
+{
+	int rc = DEVICE_GOOD;
+	char *msg = NULL;
+	struct camtape_data *softc = device;
+	struct scsi_read_attribute_values *attr_header = NULL;
+	size_t attr_size;
+	int timeout;
+	union ccb *ccb = NULL;
+
+	ltfs_profiler_add_entry(softc->profiler, NULL, TAPEBEND_REQ_ENTER(REQ_TC_WRITEATTR));
+	ltfsmsg(LTFS_DEBUG3, 30594D, "writeattr", (unsigned long long)part,
+			softc->drive_serial);
+
+	ccb = cam_getccb(softc->cd);
+	if (ccb == NULL) {
+		rc = -EDEV_NO_MEMORY;
+		goto bailout;
+	}
+
+	bzero(&(&ccb->ccb_h)[1],
+		sizeof(struct ccb_scsiio) - sizeof(struct ccb_hdr));
+
+	/*
+	 * The function interface only includes enough space for the attribute, and doesn't include
+	 * space for the header.  So we need to allocate that here, fill in the length field in the
+	 * header (which should be ignored by the target) and copy the user's data in.
+	 */
+	attr_size = size + sizeof(*attr_header);
+	attr_header = calloc(1, attr_size);
+	if (attr_header == NULL) {
+		ltfsmsg(LTFS_ERR, 10001E, "camtape_write_attribute: data buffer");
+		rc = -EDEV_NO_MEMORY;
+		goto bailout;
+	}
+	bcopy(buf, &attr_header[1], size);
+	scsi_ulto4b(size, attr_header->length);
+
+	timeout = camtape_get_timeout(softc->timeouts, WRITE_ATTRIBUTE);
+	if (timeout < 0) {
+		rc = -EDEV_UNSUPPORETD_COMMAND;
+		goto bailout;
+	}
+
+	scsi_write_attribute(&ccb->csio,
+						 /*retries*/ 1,
+						 /*cbfcnp*/ NULL,
+						 /*tag_action*/ MSG_SIMPLE_Q_TAG,
+						 /*element*/ 0,
+						 /*logical_volume*/ 0,
+						 /*partition*/ part,
+						 /*wtc*/ 1,
+						 /*data_ptr*/ (uint8_t *)attr_header,
+						 /*length*/ attr_size,
+						 /*sense_len*/ SSD_FULL_SIZE,
+						 /*timeout*/ timeout);
+
+	ccb->ccb_h.flags |= CAM_DEV_QFRZDIS | CAM_PASS_ERR_RECOVER;
+
+	rc = camtape_send_ccb(softc, ccb, &msg);
+
+	if (rc != DEVICE_GOOD)
+		camtape_process_errors(device, rc, msg, "writeattr", true);
+
+bailout:
+	if (ccb != NULL)
+		cam_freeccb(ccb);
+	if (attr_header != NULL)
+		free(attr_header);
+
+	ltfs_profiler_add_entry(softc->profiler, NULL, TAPEBEND_REQ_EXIT(REQ_TC_WRITEATTR));
+	return rc;
+}
+
+int camtape_allow_overwrite(void *device, const struct tc_position pos)
+{
+	int rc;
+	char *msg = NULL;
+	struct camtape_data *softc = device;
+	struct allow_data_overwrite append_pos;
+	int timeout;
+	union ccb *ccb;
+
+	ltfs_profiler_add_entry(softc->profiler, NULL, TAPEBEND_REQ_ENTER(REQ_TC_ALLOWOVERW));
+	ltfsmsg(LTFS_DEBUG, 3059
```

Comment:
Just for the sake of matching the coding style adopted by the project, could you remove this empty line?